### DISCUSSION TITLE
Bug or feature? `Expected type expression but received "UnionType"`

### DISCUSSION THREAD

**Discussion by ghost** [microsoft/pyright#5210] [Q&A] [answered]
### What I did:

`demo.py`
```python
Jsonish = int | str | "JsonishList"
JsonishList = list[Jsonish]
```

```sh
$ pyright demo.py
```

### What I expected:

To me this looks valid. I expect pyright to succeed without complaint.

### What actually happened:

```sh
$ pyright demo.py
./demo.py
  ./demo.py:2:20 - error: Expected type expression but received "UnionType" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
exit code: 1
```

### Context:

```sh
$ python --version
Python 3.11.3

$ pyright --version
pyright 1.1.311

$ grep pyright pyproject.toml
(no results)
```

I normally use a pretty extensive pyright config, but this reproduces with no configuration

### For contrast:

```sh
$ cat demo.py
from typing import Union

Jsonish = Union[int, str, "JsonishList"]
JsonishList = list[Jsonish]

$ pyright demo.py
0 errors, 0 warnings, 0 informations
```

The [typing docs](https://docs.python.org/3/library/stdtypes.html#union-type) assert `int | str == typing.Union[int, str]` so I'd expect these to be indistinguishable to pyright.

**Comment by erictraut** [comment, marked as answer]
Try running your sample code in the python interpreter, and you'll receive a `TypeError` exception. For this reason, pyright doesn't treat `Jsonish` as a type alias in this case; it's treated as a simple variable assignment. 

You'll want to change your code to the following:

```python
from typing import TypeAlias

Jsonish: TypeAlias = "int | str | JsonishList"
JsonishList: TypeAlias = list[Jsonish]
```

Or you can use:
```python
from typing import Union

Jsonish = Union[int, str, "JsonishList"]
JsonishList = list[Jsonish]
```

Or if you're using the latest Python 3.12 beta, you can leverage the new syntax in [PEP 695](https://peps.python.org/pep-0695/).

```python
type Jsonish = int | str | JsonishList
type JsonishList = list[Jsonish]
```

**Reply by ghost** [reply]
TL;DR: Thanks! That makes a lot of sense.

I'm in python3.11, so it doesn't TypeError (I had thought). (This particular error is unaffected by `pythonVersion = "3.11"`.)

```
$ python
Python 3.11.3 (main, Apr  7 2023, 20:13:31) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
>>> str|None
str | None
```

... but if I run the exact example from OP, then indeed I get an error, with a very helpful message. It's the forward-declaration string in particular that TypeError's. (I'd bet that in python3.20, it's treated as such.) 

```
$ python demo4.py
Traceback (most recent call last):
  File "./demo4.py", line 1, in <module>
    Jsonish = int | str | "JsonishList"
              ~~~~~~~~~~^~~~~~~~~~~~~~~
TypeError: unsupported operand type(s) for |: 'types.UnionType' and 'str'
```

From the behavior, it doesn't seem like pyright knows about this error, and gets a `UnionType` value for `Jsonish`. I recall getting a "this will error at runtime" message from pyright in some cases. I think that's appropriate here. Currently the error message is unaffected by `--pyversion 3.11`.

```
$ pyright --pythonversion 3.11 demo4.py
./demo4.py
  ./demo4.py:2:20 - error: Expected type expression but received "UnionType" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
exit code: 1
```
